src/io/XTcpSocket.cpp

```diff
@@ -0,0 +1,189 @@
+#include "XTcpSocket.h"
+
+// platform dependent system headers
+#ifdef _MSC_VER
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#define SHUT_RDWR SD_BOTH
+#define MSG_NOSIGNAL 0
+#else
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/un.h>
+#include <unistd.h>
+#endif
+
+// platform independent system headers
+#include <errno.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+void XTcpSocket::FormatLastErrorStr(const QString& func) {
+#ifdef _MSC_VER
+    int err;
+    char msgbuf[256];  // for a message up to 255 bytes.
+    msgbuf[0] = '\0';  // Microsoft doesn't guarantee this on man page.
+    err = WSAGetLastError();
+    FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,  // flags
+                  NULL,                                                        // lpsource
+                  err,                                                         // message id
+                  MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),                   // languageid
+                  msgbuf,                                                      // output buffer
+                  sizeof(msgbuf),                                              // size of msgbuf, bytes
+                  NULL);                                                       // va_list of arguments
+
+    if (!*msgbuf)
+        lastError = func + " error number : " + QString(err);  // provide error # if no string available
+    else
+        lastError = func + " error : " + msgbuf;
+
+#else
+    lastError = func + " error : " + strerror(errno);
+#endif
+}
+
+const QString XTcpSocket::getLastErrorStr(void) {
+    return lastError;
+}
+
+void XTcpSocket::close() {
+    // shutdown (sockfd,SHUT_RDWR );
+#ifdef _MSC_VER
```

Comment:
What is the reason for reverting back to these?

Reply:
The [MSVC](https://cmake.org/cmake/help/latest/variable/MSVC.html?highlight=msvc#msvc) variable is only set for cmake scripts. The [_MSC_VER](https://cmake.org/cmake/help/latest/variable/MSVC_VERSION.html#variable:MSVC_VERSION) variable is set to the preprocessor, so it can be used in the source code (as is the case).